Python/import.c

```diff
@@ -905,170 +906,169 @@ extensions_lock_release(void)
    dictionary, to avoid loading shared libraries twice.
 */
 
+static void *
+hashtable_key_from_2_strings(PyObject *str1, PyObject *str2, const char sep)
+{
+    Py_ssize_t str1_len = strlen(PyUnicode_AsUTF8(str1));
+    Py_ssize_t str2_len = strlen(PyUnicode_AsUTF8(str2));
+    assert(SIZE_MAX - str1_len - 1 - str2_len > 0);
+    size_t size = str1_len + 1 + str2_len + 1;
```

Comment:
PyUnicode_AsUTF8() _can_ fail, although it's unlikely for these strings. There's no need to call `strlen()` because the unicode object knows its length. We can assert that there's no NUL bytes in the end string, instead. The assertion doesn't make sense because SIZE_MAX is unsigned, so you're only really checking for the allocation being _exactly_ SIZE_MAX, which is actually fine (or rather, extremely likely to fail for _other_ reasons, but not because of overflow :). Since PyUnicode objects use Py_ssize_t (as returned by `PyUnicode_AsUTF8AndSize()`) we can safely assume adding two will fit in a size_t, though.

```suggestion
    Py_ssize_t str1_len, str2_len;
    const char *str1_data = PyUnicode_AsUTF8AndSize(str1, &str1_len);
    const char *str2_data = PyUnicode_AsUTF8AndSize(str2, &str2_len);
    if (str1_data == NULL || str2_data == NULL) {
        return NULL;
    }
    assert(SIZE_MAX - str1_len - str2_len > 2);
    size_t size = str1_len + 1 + str2_len + 1;
```

Reply:
> PyUnicode_AsUTF8() can fail, although it's unlikely for these strings.

For ASCII strings I don't think it can fail.  We just don't have a short-circuit code path through `PyUnicode_AsUTF8AndSize()`.

> There's no need to call strlen() because the unicode object knows its length. We can assert that there's no NUL bytes in the end string, instead.

+1

> The assertion doesn't make sense because SIZE_MAX is unsigned, so you're only really checking for the allocation being exactly SIZE_MAX, which is actually fine (or rather, extremely likely to fail for other reasons, but not because of overflow :).

My objective is to make sure we don't overflow (or try to allocate more than `size_t` bytes, since that is what `PyMem_RawMalloc()` takes).  It's a highly unlikely case, but I wanted to be careful.  If the assert would never fail then I have nothing to worry about. :)

FWIW, at first I thought we could rely on a max length for filenames and for identifiers, but it looks like the former isn't reliable and the latter doesn't exist (though dynload_shlib.c and dynload_win.c *do* effectively apply limits by truncating at large values).

> Since PyUnicode objects use Py_ssize_t (as returned by PyUnicode_AsUTF8AndSize()) we can safely assume adding two will fit in a size_t, though.

+1